### DISCUSSION TITLE
can not activate input method when set GTK_IM_MODULE=fcitx

### DISCUSSION THREAD

**Discussion by ttys3** [ghostty-org/ghostty#3628] [Feature Requests, Ideas] [answered]
## the issue

Note:
> The issue reported here is that pressing `shift` or pressing `ctrl+shift` can **NOT** activate the fcitx input method; it does not mean that fcitx cannot currently be used in Ghostty.
If I manually switch to the fcitx input method from the system tray, it can work in Ghostty.

in below description: "**works**" means pressing `shift` or pressing `ctrl+shift` CAN **activate** the fcitx input method

OS: Fedora Linux 41 Workstation Edition

DE:  GNOME 47

gtk version: 4.16.5

Window System: Wayland

related env: 

```ini
LANG=en_US.UTF-8


GTK_IM_MODULE=fcitx
QT_IM_MODULE=fcitx
XMODIFIERS=@im=fcitx
SDL_IM_MODULE=fcitx
GLFW_IM_MODULE=ibus
```

`/etc/locale.conf` content:

```
LANG="en_US.UTF-8"
```

run `env GTK_IM_MODULE=fcitx ghostty` not work, input method can not activated

in order to reproduce the issue, you need to install `fcitx5` and install a input method like `mozc`, and ensure we have `fcitx5-gtk4` (the fcitx gtk4 IM module) installed

and config fcitx5 global keys like this:

![image](https://github.com/user-attachments/assets/27c89e3c-fc5f-45ac-96d4-ecc75597d7bc)

press `shift` or press `ctrl+shift` no response, both not work. 

while in the same machine, Kitty or Gnome Terminal is OK, press `ctrl+shift`  can switch to mozc, and while in mozc we an use `shift` to switch bewteen English and non-English input

## workaround

all these workaround is almost the same method, but it is not perfect,   GTK_IM_MODULE=wayland will cause XWayland apps panic. 
 see https://github.com/ghostty-org/ghostty/discussions/3628#discussioncomment-11685434

1. using `env GTK_IM_MODULE=must-not-exists-im-module ghostty` does the tricks, and somehow, it works. but this is not the right solution

I got this tricks because Arch Linux docs said that.

> If your locale is en_US.UTF-8, and your GTK2 application cannot activate fcitx5, you can set GTK_IM_MODULE=xim [specifically for it](https://wiki.archlinux.org/title/Environment_variables#Per_application).

2. run with `unset GTK_IM_MODULE &&  ghostty` works

3. run with `env GTK_IM_MODULE=wayland ghostty` works

----------------------------

## refs docs

https://fcitx-im.org/wiki/Setup_Fcitx_5

refs:  https://wiki.archlinux.org/title/Fcitx5#GNOME

#### Wayland

Wayland's native _text-input_ protocol usually yields better results than input method modules. The Wayland frontend of fcitx5 is enabled by default, and [GTK](https://www.gtk.org)/[Qt](https://www.qt.io) utilize _text-input_ if no other IM module is explicitly specified. Therefore, it's generally recommended to only use [IM modules](#im-modules) in [Xwayland](https://wayland.freedesktop.org/xserver.html) applications. Additionally, enable [XIM](#xim) for legacy X11 applications. [More information](https://fcitx-im.org/wiki/Using_Fcitx_5_on_Wayland#TL.3BDR_Do_we_still_need_XMODIFIERS.2C_GTK_IM_MODULE_and_QT_IM_MODULE.3F).

> **Note**: _text-input_ requires the Wayland compositor and the client to use the same version of the protocol. You might want to use [IM modules](#im-modules) if the version mismatches for some applications. In particular, [GNOME](https://www.gnome.org) and [Sway](https://swaywm.org) only have support for _text-input-v3_, while [Chromium](https://www.chromium.org) defaults to _text-input-v1_ (can be overridden via `--wayland-text-input-version=3`).

<del>
## SDL (or glfw ?) related

I see that Ghostty is using `https://github.com/mitchellh/mach-glfw/`

https://github.com/slimsag/mach-glfw/commit/fb4ae48540454270ab969c8c645bbc6eff3c2dfb

> `SDL_IM_MODULE`: Set the value to fcitx. Only SDL2 requires this. SDL1 uses XIM.
> ref https://fcitx-im.org/wiki/Setup_Fcitx_5#SDL_IM_MODULE

----------------------------------------------

kitty is doing things some kind of different:

https://github.com/kovidgoyal/kitty/blob/fa6ae712a315284aadea5b13980919076014eace/glfw/ibus_glfw.c#L429

> GLFW_IM_MODULE
> This is a variable only used by [kitty](https://github.com/kovidgoyal/kitty/). You need to set it to `GLFW_IM_MODULE=ibus`
> ref https://fcitx-im.org/wiki/Setup_Fcitx_5#GLFW_IM_MODULE
</del>

**Comment by rhodes-b** [comment]
> I see that Ghostty is using https://github.com/mitchellh/mach-glfw/

Ghostty uses GLFW for developer debug builds the main runtime is GTK based so GLFW should not be in play here

**Reply by rhodes-b** [reply]
> kitty is doing things some kind of different:

> https://github.com/kovidgoyal/kitty/blob/fa6ae712a315284aadea5b13980919076014eace/glfw/ibus_glfw.c#L429

Kitty is licensed with GPL please describe what is different rather than linking the source code (Ghostty is MIT licensed so this can cause legal issues)

**Reply by ttys3** [reply]
> > kitty is doing things some kind of different:
> 
> > [kovidgoyal/kitty@`fa6ae71`/glfw/ibus_glfw.c#L429](https://github.com/kovidgoyal/kitty/blob/fa6ae712a315284aadea5b13980919076014eace/glfw/ibus_glfw.c#L429)
> 
> Kitty is licensed with GPL please describe what is different rather than linking the source code (Ghostty is MIT licensed so this can cause legal issues)

I do not know. just see Kitty is doing strange things. it trigger input method like fcitx5 (fcitx5 it self is also a input method framework just the same level as ibus) via ibus  (input framework) dbus. I do not think Ghostty need this. (since the main runtime is GTK based)

**Comment by ttys3** [comment]
I am not familiar with Ghostty's code, nor do I understand zig. I am just a user trying to solve the input method problem. At the same time, if possible, I would like to provide as much information and help as I can.

What I want to ask now is, for Ghosttty, do I need to set the GTK_IM_MODULE or GLFW_IM_MODULE variable?

as you answered, I think `GLFW_IM_MODULE` or glfw is not related to the input method issue.

**Comment by ttys3** [comment]
it is very strange. Ghosttty can activate input method while `GTK_IM_MODULE` is set to any IM module that does NOT exists.

```
env GTK_IM_MODULE=must-not-exists-im-module ghostty


(process:855748): Gtk-WARNING **: 11:44:41.760: No IM module matching GTK_IM_MODULE=must-not-exists-im-module found
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60 NVIDIA
```

set to `GTK_IM_MODULE=xim` works is just by accident:

```
(process:856607): Gtk-WARNING **: 11:46:41.065: No IM module matching GTK_IM_MODULE=xim found
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60 NVIDIA
```

related gtk code: https://github.com/GNOME/gtk/blob/e95a7c5dfc6dd672fbefe1c8951803801e743416/gtk/gtkimmodule.c#L187

https://github.com/GNOME/gtk/blob/e95a7c5dfc6dd672fbefe1c8951803801e743416/gtk/gtkimmodule.c#L177C1-L247C2

**Comment by mitchellh** [comment]
This feels like a GTK bug to me, especially based on the other comments.

**Reply by ttys3** [reply]
> This feels like a GTK bug to me, especially based on the other comments.

I do not know if it is, maybe fire a bug on GNOME.

but other Terminals works well with the env setting `GTK_IM_MODULE=fcitx`

like 

```
env GTK_IM_MODULE=fcitx && alacritty

env GTK_IM_MODULE=fcitx && gnome-terminal

env GTK_IM_MODULE=fcitx && ptyxis
```

**Reply by mitchellh** [reply]
The only interesting one here is `ptyxis` because I think it's also GTK4. @tristan957 do you have any knowledge on this topic?

Fwiw we don't do anything special with input methods. We attach standard event handlers to input method events (generic) but don't try to iterate, set, unset, etc. input methods.

**Reply by tristan957** [reply]
I know nothing about input modules, so I cannot comment, but I haven't seen any Ghostty code that deals with them.

**Comment by ttys3** [comment]
after do `fcitx5-diagnose` , It reminded me:

## Gtk:
1.  gtk - `${GTK_IM_MODULE}`:

    Environment variable GTK_IM_MODULE is set to "fcitx" correctly.

    Using fcitx5-gtk2-immodule-probing to check the actual im module to be used under current environment:

        GTK_IM_MODULE=fcitx

    Using fcitx5-gtk3-immodule-probing to check the actual im module to be used under current environment:

        GTK_IM_MODULE=fcitx

    Using fcitx5-gtk4-immodule-probing to check the actual im module to be used under current environment:

        GTK_IM_MODULE=fcitx


So, I went ahead and conducted the following tests:

```shell
❯ fcitx5-gtk4-immodule-probing
GTK_IM_MODULE=fcitx
```

```shell
❯ env GTK_IM_MODULE=xxxx fcitx5-gtk4-immodule-probing

(process:890674): Gtk-WARNING **: 12:42:25.597: No IM module matching GTK_IM_MODULE=xxxx found
GTK_IM_MODULE=wayland
```

```shell
 ❯ unset GTK_IM_MODULE && fcitx5-gtk4-immodule-probing
GTK_IM_MODULE=wayland
```

so we got the correct env value here.

with `env GTK_IM_MODULE=wayland ghostty`, now the input method can be activated

**Comment by ttys3** [comment]
@mitchellh I think we still have problems when run with `env GTK_IM_MODULE=wayland && ghostty`

the problem is, when user run a non-wayland native app, the app it self still need `GTK_IM_MODULE=fcitx`,  and `GTK_IM_MODULE=wayland` will cause the app panic.

for example, running vscode from Ghostty, since vscode currently default to X11 mode, so it runs as XWayland client

**Reply by mitchellh** [reply]
Workaround for now would be to probably unset this (or reset it) in your shell.

I'm just not sure if Ghostty should be setting this or not on its own...

**Reply by ttys3** [reply]
I think Ghostty should not touch these environments.

the problem is why it can not play well with `GTK_IM_MODULE=fcitx`, while other terminals do works.

I'm not really knowledgeable about input method programming, so I can't offer much help. If I could, I'd love to do something to assist.

**Reply by mitchellh** [reply]
> while other terminals do works.

Again, the only one that matters here is probably `ptyxis`. The others aren't GTK based so they're not comparable. I suspect that there is a GTK issue here still, as I originally said.

**Reply by ttys3** [reply]
> > while other terminals do works.
> 
> Again, the only one that matters here is probably `ptyxis`. The others aren't GTK based so they're not comparable. I suspect that there is a GTK issue here still, as I originally said.

I do not think it has relation to ptyxis itself, since it uses [libvte](https://gitlab.gnome.org/GNOME/vte.git), just like gnome terminal.


seems it calls many `gtk_im_context_xxxxxxxxxxxx` functions, like :

https://gitlab.gnome.org/GNOME/vte/-/blob/a5ca69e49b6ad7020d997e7125f15ce55279ef0a/src/widget.cc?page=2#L1911

```cpp
void
Widget::realize() noexcept
{
        //        m_mouse_cursor_over_widget = false;  /* We'll receive an enter_notify_event if the window appears under the cursor. */

	/* Create stock cursors */
	m_default_cursor = create_cursor(VTE_DEFAULT_CURSOR);
	m_invisible_cursor = create_cursor("none"s);
	m_mousing_cursor = create_cursor(VTE_MOUSING_CURSOR);
        if (_vte_debug_on(VTE_DEBUG_HYPERLINK))
                /* Differ from the standard regex match cursor in debug mode. */
                m_hyperlink_cursor = create_cursor(VTE_HYPERLINK_CURSOR_DEBUG);
        else
                m_hyperlink_cursor = create_cursor(VTE_HYPERLINK_CURSOR);

#if VTE_GTK == 3
	/* Create an input window for the widget. */
        auto allocation = m_terminal->get_allocated_rect();
	GdkWindowAttr attributes;
	attributes.window_type = GDK_WINDOW_CHILD;
	attributes.x = allocation.x;
	attributes.y = allocation.y;
	attributes.width = allocation.width;
	attributes.height = allocation.height;
	attributes.wclass = GDK_INPUT_ONLY;
	attributes.visual = gtk_widget_get_visual(m_widget);
	attributes.event_mask =
                gtk_widget_get_events(m_widget) |
                GDK_EXPOSURE_MASK |
                GDK_FOCUS_CHANGE_MASK |
                GDK_SMOOTH_SCROLL_MASK |
                GDK_SCROLL_MASK |
                GDK_BUTTON_PRESS_MASK |
                GDK_BUTTON_RELEASE_MASK |
                GDK_POINTER_MOTION_MASK |
                GDK_BUTTON1_MOTION_MASK |
                GDK_ENTER_NOTIFY_MASK |
                GDK_LEAVE_NOTIFY_MASK |
                GDK_KEY_PRESS_MASK |
                GDK_KEY_RELEASE_MASK;
	attributes.cursor = m_default_cursor.get();
	guint attributes_mask =
                GDK_WA_X |
                GDK_WA_Y |
                (attributes.visual ? GDK_WA_VISUAL : 0) |
                GDK_WA_CURSOR;

	m_event_window = gdk_window_new(gtk_widget_get_parent_window (m_widget),
                                        &attributes, attributes_mask);
        gtk_widget_register_window(m_widget, m_event_window);
#endif /* VTE_GTK == 3 */

        assert(!m_im_context);
        m_im_context = vte::glib::take_ref(gtk_im_multicontext_new());
#if (VTE_GTK == 3 && GTK_CHECK_VERSION (3, 24, 14)) || VTE_GTK == 4
        g_object_set(m_im_context.get(),
                     "input-purpose", GTK_INPUT_PURPOSE_TERMINAL,
                     nullptr);
#endif

#if VTE_GTK == 3
	gtk_im_context_set_client_window(m_im_context.get(), m_event_window);
#elif VTE_GTK == 4
        gtk_im_context_set_client_widget(m_im_context.get(), gtk());
#endif
	g_signal_connect(m_im_context.get(), "commit",
			 G_CALLBACK(im_commit_cb), this);
	g_signal_connect(m_im_context.get(), "preedit-start",
			 G_CALLBACK(im_preedit_start_cb), this);
	g_signal_connect(m_im_context.get(), "preedit-changed",
			 G_CALLBACK(im_preedit_changed_cb), this);
	g_signal_connect(m_im_context.get(), "preedit-end",
			 G_CALLBACK(im_preedit_end_cb), this);
	g_signal_connect(m_im_context.get(), "retrieve-surrounding",
			 G_CALLBACK(im_retrieve_surrounding_cb), this);
	g_signal_connect(m_im_context.get(), "delete-surrounding",
			 G_CALLBACK(im_delete_surrounding_cb), this);
	gtk_im_context_set_use_preedit(m_im_context.get(), true);

        m_clipboard = std::make_shared<Clipboard>(*this, ClipboardType::CLIPBOARD);
        m_primary_clipboard = std::make_shared<Clipboard>(*this, ClipboardType::PRIMARY);

        m_terminal->widget_realize();
}
```

**Reply by rhodes-b** [reply]
Please delete this, vte is GPLv3 which can cause issues with Ghostty as its licensed as MIT

**Reply by rhodes-b** [reply]
@ttys3 ^ Very stongly suggesed

**Reply by mitchellh** [reply]
I hid the comment

**Comment by ttys3** [comment]
I reported an [issue](https://github.com/fcitx/fcitx5/issues/1218) on fcitx; it might be related.

**Comment by wengxt** [comment]
Hi fcitx dev here. 

I believe there's a bug between gtk's wayland im module  and ghostty. One thing to notice that on Wayland, filter_key is not used. The filter is done on the compoistor side. 

The patch here https://github.com/ghostty-org/ghostty/discussions/3279#discussioncomment-11679607:

```
-    if (!self.in_keypress) return;
+    // if (!self.in_keypress) return;
```

I think this is the key change, in_keypress will not be true if input method consume the key.

If you want to learn more about the difference when text-input protocol is used, I have a blog post talked about this:

https://www.csslayer.info/wordpress/linux/key-repetition-and-key-event-handling-issue-with-wayland-input-method-protocols/ 

No need to bother about the actual issue the blog is complaining about, just focus on the the first two graph about the message interaction. also while the input method <-> compositor protocol are different among different compositors (mutter uses ibus dbus itnerface, kwin uses zwp_input_method_v1, wlroots uses zwp_input_method_v2 ), the model is same and the different is not visible to application since application uses text-input to talk to compositor.

TL;DR:
* When wayland native protocol (text-input) is used, Key event flows like: kernel -> compositor -> input method -> application (gtk_im_context_filter_key does nothing).
* When the old model (whatever private im module, or xim): Key event flows like: kernel -> display server -> application ---(via gtk_im_context_filter_keypress in ghostty)----> input method -> Back to application

**Reply by ttys3** [reply]
@wengxt @mitchellh yes, that patch fixed the Preedit not shown up issue. but not related to mine.

I'm not saying I can't use the fcitx input method when `GTK_IM_MODULE=fcitx`; I can always use it if I manually switch the input method from the system tray.

However, when the default input method is set to `English` in fcitx, I can't activate it using ctrl+shift or shift in Ghostty.

**Reply by wengxt** [reply]
@ttys3 it's somewhat related. After preedit is fixed, I believe GTK_IM_MODULE=wayland doesn't have the modifier only hotkey not working issue. Because for text-input model, the key always goes to input method before application. So it can be a workaround for your case.

@mitchellh As for "Ctrl" or "Ctrl+Shift" not working in ghostty, the issue I observed from fcitx side is that, the key release is not forwarded to input method. 

Since you mentioned you kind of know how to test fcitx5, you can use following command

```
dbus-send --type=method_call --print-reply=literal --dest=org.fcitx.Fcitx5 /controller org.fcitx.Fcitx.Controller1.SetLogRule 'string:key_trace=5'
```

to enable fcitx verbose key logging at runtime

This is what happens in ghostty when you press left shift repeatedly:
```
inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=2147483648) origKey: Key(Shift_L states=2147483648) Release:0 keycode: 50
inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=2147483648) origKey: Key(Shift_L states=2147483648) Release:0 keycode: 50
inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=2147483648) origKey: Key(Shift_L states=2147483648) Release:0 keycode: 50
inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=2147483648) origKey: Key(Shift_L states=2147483648) Release:0 keycode: 50
inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=2147483648) origKey: Key(Shift_L states=2147483648) Release:0 keycode: 50
```

while in other gtk4 app:
```
D2024-12-28 18:19:48.901382 instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=0) origKey: Key(Shift_L states=0) Release:0 keycode: 50
D2024-12-28 18:19:48.901929 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-28 18:19:49.026939 instance.cpp:928] KeyEvent: Key(Shift+Shift_L states=1) rawKey: Key(Shift+Shift_L states=1) origKey: Key(Shift+Shift_L states=1) Release:1 keycode: 50
D2024-12-28 18:19:49.027370 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-28 18:19:49.129464 instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=0) origKey: Key(Shift_L states=0) Release:0 keycode: 50
D2024-12-28 18:19:49.130018 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-28 18:19:49.234842 instance.cpp:928] KeyEvent: Key(Shift+Shift_L states=1) rawKey: Key(Shift+Shift_L states=1) origKey: Key(Shift+Shift_L states=1) Release:1 keycode: 50
D2024-12-28 18:19:49.235273 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-28 18:19:49.343543 instance.cpp:928] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=0) origKey: Key(Shift_L states=0) Release:0 keycode: 50
D2024-12-28 18:19:49.344153 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-28 18:19:49.445169 instance.cpp:928] KeyEvent: Key(Shift+Shift_L states=1) rawKey: Key(Shift+Shift_L states=1) origKey: Key(Shift+Shift_L states=1) Release:1 keycode: 50
```

As you can see, the difference is the "Shift+Shift_L" release is missing. I believe it is ghostty filtered it away before gtk_im_context_filter_keypress.

In this case, I think this code is fishy:

https://github.com/ghostty-org/ghostty/blob/5293fc9c2f19b7e01472de8c2ef08bef603b754a/src/apprt/gtk/Surface.zig#L1555

This comment is wrong. All key event need to be forwarded to input method, including release. Many input method features relies on key release to work. In this case, modifier only hotkey can only be triggered on key release event ( you can imagine why key press doesn't work).

I can name quite a few other features that relies on key release event, including anthy's nicola layout, rime's combo-pinyin mode.
In short, both key press & key release events need to be forwarded to input method.

**Reply by ttys3** [reply]
> @ttys3 it's somewhat related. After preedit is fixed, I believe GTK_IM_MODULE=wayland doesn't have the modifier only hotkey not working issue. 

GTK_IM_MODULE=wayland always works with or without the preedit issue fixed.  the hotkey trigger can always activate fcitx.

> Because for text-input model, the key always goes to input method before application. So it can be a workaround for your case.

Thanks for clarifying. So, GTK_IM_MODULE=wayland works because it operates independently of the application's key event handling, correct?

**Reply by ttys3** [reply]
yes, I see the difference, I tested it and got the same result.  Ghostty did not send any release events to fcitx.

```shell
#fcitx5 -D -r --verbose 'default=4,key_trace=5'

dbus-send --type=method_call --print-reply=literal --dest=org.fcitx.Fcitx5 /controller org.fcitx.Fcitx.Controller1.SetLogRule 'string:key_trace=5'
```
## Press `Ctrl+Shift`

Ghostty:

```shell
D2024-12-29 04:37:09.279231 instance.cpp:919] KeyEvent: Key(Control_L states=0) rawKey: Key(Control_L states=0) origKey: Key(Control_L states=0) Release:0 keycode: 37
D2024-12-29 04:37:09.279281 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:37:09.279482 instance.cpp:919] KeyEvent: Key(Control_L states=0) rawKey: Key(Control_L states=2147483648) origKey: Key(Control_L states=2147483648) Release:0 keycode: 37
D2024-12-29 04:37:09.279494 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:37:09.319266 instance.cpp:919] KeyEvent: Key(Control+Shift_L states=4) rawKey: Key(Control+Shift_L states=4) origKey: Key(Control+Shift_L states=4) Release:0 keycode: 50
D2024-12-29 04:37:09.319293 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:37:09.319585 instance.cpp:919] KeyEvent: Key(Control+Shift_L states=4) rawKey: Key(Control+Shift_L states=2147483652) origKey: Key(Control+Shift_L states=2147483652) Release:0 keycode: 50
D2024-12-29 04:37:09.319621 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
```

Ptyxis:
```shell
D2024-12-29 04:40:24.094258 instance.cpp:919] KeyEvent: Key(Control_L states=0) rawKey: Key(Control_L states=0) origKey: Key(Control_L states=0) Release:0 keycode: 37
D2024-12-29 04:40:24.094317 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:40:24.102105 instance.cpp:919] KeyEvent: Key(Control+Shift_L states=4) rawKey: Key(Control+Shift_L states=4) origKey: Key(Control+Shift_L states=4) Release:0 keycode: 50
D2024-12-29 04:40:24.102136 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:40:24.182065 instance.cpp:919] KeyEvent: Key(Control+Shift+Shift_L states=5) rawKey: Key(Control+Shift+Shift_L states=5) origKey: Key(Control+Shift+Shift_L states=5) Release:1 keycode: 50
D2024-12-29 04:40:24.182195 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:40:24.205386 instance.cpp:919] KeyEvent: Key(Control+Control_L states=4) rawKey: Key(Control+Control_L states=4) origKey: Key(Control+Control_L states=4) Release:1 keycode: 37
D2024-12-29 04:40:24.205421 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
```


## Press `Shift` only

Ghostty:
```shell
D2024-12-29 04:38:18.015586 instance.cpp:919] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=0) origKey: Key(Shift_L states=0) Release:0 keycode: 50
D2024-12-29 04:38:18.015629 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:38:18.015939 instance.cpp:919] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=2147483648) origKey: Key(Shift_L states=2147483648) Release:0 keycode: 50
D2024-12-29 04:38:18.015960 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
```

Ptyxis:
```shell
D2024-12-29 04:39:02.086717 instance.cpp:919] KeyEvent: Key(Shift_L states=0) rawKey: Key(Shift_L states=0) origKey: Key(Shift_L states=0) Release:0 keycode: 50
D2024-12-29 04:39:02.086771 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
D2024-12-29 04:39:02.151669 instance.cpp:919] KeyEvent: Key(Shift+Shift_L states=1) rawKey: Key(Shift+Shift_L states=1) origKey: Key(Shift+Shift_L states=1) Release:1 keycode: 50
D2024-12-29 04:39:02.151692 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
```

**Comment by ttys3** [comment]
@mitchellh As we communicated with the fcitx developer, there might still be some issues with the implementation on Ghostty's side. I think we should convert this discussion into an issue.

**Reply by mitchellh** [reply]
Until the issue is identified, I will keep it a discussion. I'm actively watching these. I agree with you there's improvements Ghostty needs to make here but need a better understanding on my end before opening the issue.

**Reply by wengxt** [reply]
@mitchellh I think the issues are pretty clear (even with code pointers).

1. key release event also need to forward to input method. This is not the first time I saw such issue in widget/toolkit implementation. 
https://github.com/libsdl-org/SDL/pull/5281 is one of the example of such issue on SDL that I fixed in the past.

2. preedit handling should not be associated with in in_filter_key, because on wayland the key filter is on display server side, such check does not really make sense. Also imagine if it is not a keyboard-based input method, but an handwriting input method, or even on-screen keyboard, the filter "key" will never be triggered, but engine may still send preedit update.

**Comment by hooklab** [comment]
Same problem with me

**Comment by hooklab** [comment]
```
run with unset GTK_IM_MODULE &&  ghostty
run with env GTK_IM_MODULE=wayland && ghostty
```

I've tried the above and still can't switch

**Reply by ttys3** [reply]
there's a typo, it is `env GTK_IM_MODULE=wayland ghostty`

**Reply by hooklab** [reply]
> there's a typo, it is `env GTK_IM_MODULE=wayland ghostty`

the same

**Comment by mitchellh** [comment, marked as answer]
Thank you for all the input. I made an issue #4332 to track this and link to the critical responses.
